app/validators/sections/section_validator.py

```diff
@@ -40,23 +43,65 @@ def validate_list_exists(self, list_name):
         if list_name not in self.questionnaire_schema.list_names:
             self.add_error(error_messages.FOR_LIST_NEVER_POPULATED, list_name=list_name)
 
+    def validate_skip_conditions(self, skip_conditions, origin_id):
+        for skip_condition in skip_conditions:
+            when_validator = WhenRuleValidator(
+                skip_condition["when"], origin_id, self.questionnaire_schema
+            )
+            self.errors += when_validator.validate()
+
+    def validate_value_sources(self):
+        source_references = get_object_containing_key(self.section, "identifier")
+        for json_path, source_reference in source_references:
+            if "source" in source_reference:
+                value_source_validator = ValueSourceValidator(
+                    value_source=source_reference,
+                    json_path=json_path,
+                    questionnaire_schema=self.questionnaire_schema,
+                )
+                self.errors += value_source_validator.validate()
+
     def validate_groups(self):
```

Comment:
This could be refactored as it shares the same logic as the code underneath, we can add a new method like this:
```suggestion
    def validate_groups(self):
        for group in self.section["groups"]:
            self.validate_unit(group, group)
            self.validate_blocks(group["id"])

    def validate_blocks(self, group_id):
        group = self.questionnaire_schema.get_group(group_id)

        for block in group.get("blocks"):
            self.validate_unit(block, group)

            block_validator = get_block_validator(block, self.questionnaire_schema)
            self.errors += block_validator.validate()

            self.validate_question(block)
            self.validate_variants(block)

    def validate_unit(self, unit, group):
        if "routing_rules" in unit:
            if any("goto" in rule for rule in unit["routing_rules"]):
                block_routing_validator = RoutingValidator(
                    unit, group, self.questionnaire_schema
                )
            else:
                block_routing_validator = NewRoutingValidator(
                    routing_rules=unit["routing_rules"],
                    group=group,
                    origin_id=unit["id"],
                    questionnaire_schema=self.questionnaire_schema,
                )
            self.errors += block_routing_validator.validate()

        if "skip_conditions" in unit:
            self.validate_skip_conditions(unit["skip_conditions"], unit["id"])

```

Reply:
Good shout. I have called the common method `validate_routing` and used `schema_element` in place of `unit` in https://github.com/ONSdigital/eq-questionnaire-validator/pull/97/commits/41eaf12b93a601c7ebb325e28d2627890aa53458

Reply:
Thanks👍